FWCore/SharedMemory/interface/ControllerChannel.h

```diff
@@ -0,0 +1,132 @@
+#ifndef FWCore_SharedMemory_ControllerChannel_h
+#define FWCore_SharedMemory_ControllerChannel_h
+// -*- C++ -*-
+//
+// Package:     FWCore/SharedMemory
+// Class  :     ControllerChannel
+//
+/**\class ControllerChannel ControllerChannel.h " FWCore/SharedMemory/interface/ControllerChannel.h"
+
+ Description: Primary communication channel for the Controller process
+
+ Usage:
+    Works in conjunction with the WorkerChannel
+
+*/
+//
+// Original Author:  Chris Jones
+//         Created:  21/01/2020
+//
+
+// system include files
+#include <string>
+#include <iostream>
+#include "boost/interprocess/managed_shared_memory.hpp"
+#include "boost/interprocess/sync/named_mutex.hpp"
+#include "boost/interprocess/sync/named_condition.hpp"
+#include "boost/interprocess/sync/scoped_lock.hpp"
+
+// user include files
+#include "FWCore/Utilities/interface/Transition.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+// forward declarations
+
+namespace edm::shared_memory {
+  class ControllerChannel {
+  public:
+    /** iName is used as the base for the shared memory name. The full name uses iID as well as getpid() to create the value sharedMemoryName().
+     iID allows multiple ControllChannels to use the same base name iName.
+     */
+    ControllerChannel(std::string const& iName, int iID);
+    ~ControllerChannel();
+    ControllerChannel(const ControllerChannel&) = delete;
+    const ControllerChannel& operator=(const ControllerChannel&) = delete;
+
+    // ---------- member functions ---------------------------
+
+    /** setupWorker must be called only once and done before any calls to doTransition. The functor iF should setup values associated
+     with shared memory use, such as manipulating the value from toWorkerBufferIndex(). The call to setupWorker proper synchronizes
+     the Controller and Worker processes.
+     */
+    template <typename F>
+    void setupWorker(F&& iF) {
+      using namespace boost::interprocess;
+      scoped_lock<named_mutex> lock(mutex_);
+      iF();
+      using namespace boost::posix_time;
+      //std::cout << id_ << " waiting for external process" << std::endl;
```

Comment:
Are the commented-out `cout`s still useful?

Reply:
They were extremely useful when diagnosing problems. I'd like to keep them until at least I try to build the generator version.

Reply:
On a further thought they look useful for any future debugging as well. I presume you did not want to add a dependence on MessageLogger?

Reply:
Not at that time. It would also make debugging using the stand alone program more difficult.